test/common/common/utility_test.cc

```diff
@@ -116,6 +118,65 @@ TEST(DateUtil, NowToMilliseconds) {
   EXPECT_EQ(12345067, DateUtil::nowToMilliseconds(test_time));
 }
 
+TEST(OutputBufferStream, FailsOnWriteToEmptyBuffer) {
+  constexpr char data = 'x';
+  OutputBufferStream ostream{nullptr, 0};
+  ASSERT_TRUE(ostream.good());
+
+  ostream << data;
+
+  EXPECT_TRUE(ostream.bad());
+}
+
+TEST(OutputBufferStream, CanWriteToBuffer) {
+  constexpr char data[] = "123";
+  std::array<char, 3> buffer;
+
+  OutputBufferStream ostream{buffer.data(), buffer.size()};
+  ostream << data;
+
+  EXPECT_EQ(std::string(buffer.data(), ostream.numValidBytesWrittenToBuffer()), data);
```

Comment:
nit: consider using absl::string_view instead of std::string, but also see comments in source/common/common/utility.h